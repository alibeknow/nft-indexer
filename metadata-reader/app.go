package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
