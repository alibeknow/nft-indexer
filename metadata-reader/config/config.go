package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
