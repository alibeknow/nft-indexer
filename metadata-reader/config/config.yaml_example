server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
