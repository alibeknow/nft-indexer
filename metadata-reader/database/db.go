package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
