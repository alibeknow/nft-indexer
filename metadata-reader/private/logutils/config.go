package logutils

type Config struct {
	Level string
}
