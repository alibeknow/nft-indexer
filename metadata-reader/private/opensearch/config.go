package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
