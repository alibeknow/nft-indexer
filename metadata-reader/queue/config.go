package queue

import "fmt"

type Config struct {
	Protocol string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	Workers  int
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	return url
}
