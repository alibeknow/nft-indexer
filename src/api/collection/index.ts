export * from './collection.controller';
export * from './collection.module';
export * from './collection.service';
