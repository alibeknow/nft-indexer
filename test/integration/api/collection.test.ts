import { CollectionModule } from '@api/collection';
import { HttpServer, INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { Blockchain } from '@shared/blockchain';
import { ContractsRepo } from '@shared/contracts';
import { DB, DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { TokenStandard } from '@shared/tokens';
import { UpdateResult } from 'mongodb';
import { LoggerModule } from 'nestjs-pino';
import nock from 'nock';
import request from 'supertest';
import { getNestApp } from '../../helpers/app';
import { staxResponseFixture } from './fixtures';
import { ConfigModule } from '@nestjs/config';
import { getApiConfig } from '@api/app.config';
import { openseaResponseFixture } from './fixtures/opensea-response.fixture';

const collection = openseaResponseFixture.collection;

nock('https://test.opensea/').persist().get(/.*/).reply(200, { collection });

async function seedContractsRepo(db: DB): Promise<UpdateResult[]> {
  const contractsRepo = new ContractsRepo(db);

  const contractsList = [
    {
      blockchain: Blockchain.ETH,
      address: '0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f',
      type: TokenStandard.ERC1155,
      name: 'Superlative Mutated Apes',
      block: 14316461,
      createdAt: new Date(),
    },
    {
      blockchain: Blockchain.ETH,
      address: '0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c',
      type: TokenStandard.ERC721,
      name: 'Audiovisions',
      block: 14316461,
      createdAt: new Date(),
    },
    {
      blockchain: Blockchain.ETH,
      address: '0x83d062769d78deD9B051E69c80fA6CC091dbF27B',
      type: TokenStandard.ERC721,
      name: 'MavionWorld',
      block: 14316467,
      createdAt: new Date(),
    },
  ];

  return Promise.all(
    contractsList.map((item) => contractsRepo.insertIfNone(item)),
  );
}

jest.setTimeout(50000);

describe('[Collection endpoint]', () => {
  let app: INestApplication;
  let httpServer: HttpServer;
  let db: DB;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `api-collection-test-${dbConfig.db}` });
    await db.open();

    const moduleRef = await Test.createTestingModule({
      imports: [
        CollectionModule,
        LoggerModule.forRoot(),
        ConfigModule.forRoot({
          load: [ getApiConfig ],
          isGlobal: true,
        }),
      ],
    })
      .overrideProvider(DB)
      .useValue(db)
      .compile();

    app = await getNestApp(moduleRef);
    httpServer = app.getHttpServer();

    nock('https://test.stax')
      .persist()
      .get(/.*/)
      .reply(200, { staxResponseFixture });
  });

  describe('[GET /api/v0/:chainName/contracts/:contractId]', () => {
    it('should return 404 error if requested item doesn`t exist in our DB', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v0/eth/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(404);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return 400 error if validation doesn`t pass', async () => {
      const actualResult = await request(httpServer)
        .get(
          '/api/v0/solana/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73',
        )
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct results based on params', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v0/eth/contracts/0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v0/eth/contracts/0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v0/eth/contracts/0x83d062769d78deD9B051E69c80fA6CC091dbF27B')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();
      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });

    it('should return correct results even if contract address is not checksum formatted', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v0/eth/contracts/0x9fb2eeb75754815c5cc9092cd53549cea5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v0/eth/contracts/0X20Cc3Bd22144209CF93dA14017C506b0ed0F3e4C')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v0/eth/contracts/0x83d062769d78ded9b051e69c80fa6cc091dbf27b')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();

      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();

      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });
  });

  describe('[GET /api/v1/:chainName/contracts/:contractId]', () => {
    it('should return 404 error if requested item doesn`t exist in our DB', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v1/eth/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(404);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return 400 error if validation doesn`t pass', async () => {
      const actualResult = await request(httpServer)
        .get(
          '/api/v1/solana/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73',
        )
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct results based on params', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v1/eth/contracts/0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v1/eth/contracts/0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v1/eth/contracts/0x83d062769d78deD9B051E69c80fA6CC091dbF27B')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();
      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });

    it('should return correct results even if contract address is not checksum formatted', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v1/eth/contracts/0x9fb2eeb75754815c5cc9092cd53549cea5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v1/eth/contracts/0X20Cc3Bd22144209CF93dA14017C506b0ed0F3e4C')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v1/eth/contracts/0x83d062769d78ded9b051e69c80fa6cc091dbf27b')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();

      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();

      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });
  });

  describe('[GET /api/v2/:chainName/contracts/:contractId]', () => {
    it('should return 404 error if requested item doesn`t exist in our DB', async () => {
      const actualResult = await request(httpServer)
        .get('/api/v2/eth/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73')
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(404);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return 400 error if validation doesn`t pass', async () => {
      const actualResult = await request(httpServer)
        .get(
          '/api/v2/solana/contracts/0x33EeCbf908478C10614626A9D304bfe18B78DD73',
        )
        .set('Accept', 'application/json');

      expect(actualResult.statusCode).toBe(400);
      expect(actualResult.body).toMatchSnapshot();
    });

    it('should return correct results based on params', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v2/eth/contracts/0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v2/eth/contracts/0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v2/eth/contracts/0x83d062769d78deD9B051E69c80fA6CC091dbF27B')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();
      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();
      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });

    it('should return correct results even if contract address is not checksum formatted', async () => {
      await seedContractsRepo(db);

      const actualResult1 = await request(httpServer)
        .get('/api/v2/eth/contracts/0x9fb2eeb75754815c5cc9092cd53549cea5dc404f')
        .set('Accept', 'application/json');

      const actualResult2 = await request(httpServer)
        .get('/api/v2/eth/contracts/0X20Cc3Bd22144209CF93dA14017C506b0ed0F3e4C')
        .set('Accept', 'application/json');

      const actualResult3 = await request(httpServer)
        .get('/api/v2/eth/contracts/0x83d062769d78ded9b051e69c80fa6cc091dbf27b')
        .set('Accept', 'application/json');

      expect(actualResult1.statusCode).toBe(200);
      expect(actualResult1.body).toMatchSnapshot();

      expect(actualResult2.statusCode).toBe(200);
      expect(actualResult2.body).toMatchSnapshot();

      expect(actualResult3.statusCode).toBe(200);
      expect(actualResult3.body).toMatchSnapshot();
    });
  });

  afterEach(async () => {
    await db.contracts().deleteMany({});
  });

  afterAll(async () => {
    nock.cleanAll();
    await app.close();
  });
});
