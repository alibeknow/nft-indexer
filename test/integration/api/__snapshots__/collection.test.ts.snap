// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return 400 error if validation doesn\`t pass 1`] = `
{
  "error": "Bad Request",
  "message": [
    "chainName must be a valid enum value",
  ],
  "statusCode": 400,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return 404 error if requested item doesn\`t exist in our DB 1`] = `
{
  "message": "Not Found",
  "statusCode": 404,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results based on params 1`] = `
{
  "result": {
    "id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "name": "Superlative Mutated Apes",
    "type": "ERC-1155",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results based on params 2`] = `
{
  "result": {
    "id": "eth:0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "name": "Audiovisions",
    "type": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results based on params 3`] = `
{
  "result": {
    "id": "eth:0x83d062769d78deD9B051E69c80fA6CC091dbF27B",
    "name": "MavionWorld",
    "type": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 1`] = `
{
  "result": {
    "id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "name": "Superlative Mutated Apes",
    "type": "ERC-1155",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 2`] = `
{
  "result": {
    "id": "eth:0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "name": "Audiovisions",
    "type": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v0/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 3`] = `
{
  "result": {
    "id": "eth:0x83d062769d78deD9B051E69c80fA6CC091dbF27B",
    "name": "MavionWorld",
    "type": "ERC-721",
  },
  "statusCode": 200,
}
`;

// V1

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return 400 error if validation doesn\`t pass 1`] = `
{
  "error": "Bad Request",
  "message": [
    "chainName must be a valid enum value",
  ],
  "statusCode": 400,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return 404 error if requested item doesn\`t exist in our DB 1`] = `
{
  "message": "No metadata found for the contract address 0x33EeCbf908478C10614626A9D304bfe18B78DD73",
  "statusCode": 404,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results based on params 1`] = `
{
  "result": {
    "address": "0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "name": "Superlative Mutated Apes",
    "tokenType": "ERC-1155",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results based on params 2`] = `
{
  "result": {
    "address": "0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "name": "Audiovisions",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results based on params 3`] = `
{
  "result": {
    "address": "0x83d062769d78deD9B051E69c80fA6CC091dbF27B",
    "name": "MavionWorld",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 1`] = `
{
  "result": {
    "address": "0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "name": "Superlative Mutated Apes",
    "tokenType": "ERC-1155",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 2`] = `
{
  "result": {
    "address": "0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "name": "Audiovisions",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v1/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 3`] = `
{
  "result": {
    "address": "0x83d062769d78deD9B051E69c80fA6CC091dbF27B",
    "name": "MavionWorld",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;

// V2

exports[`[Collection endpoint] [GET /api/v2/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 1`] = `
{
  "result": {
    "collectionImage": "some image",
    "contractAddress": "0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f",
    "name": "Superlative Mutated Apes",
    "tokenType": "ERC-1155",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v2/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 2`] = `
{
  "result": {
    "collectionImage": "some image",
    "contractAddress": "0x20Cc3Bd22144209CF93dA14017C506b0ed0F3e4c",
    "name": "Audiovisions",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;

exports[`[Collection endpoint] [GET /api/v2/:chainName/contracts/:contractId] should return correct results even if contract address is not checksum formatted 3`] = `
{
  "result": {
    "collectionImage": "some image",
    "contractAddress": "0x83d062769d78deD9B051E69c80fA6CC091dbF27B",
    "name": "MavionWorld",
    "tokenType": "ERC-721",
  },
  "statusCode": 200,
}
`;
